velox/common/memory/Memory.h

```diff
@@ -67,87 +67,90 @@ constexpr uint16_t kDefaultAlignment = 64;
+  /// Called by MemoryManager and MemoryPool upon memory usage updates and
+  /// propagates down the subtree recursively.
   virtual void capMemoryAllocation() = 0;
-  // Called by MemoryManager and propagates down recursively if applicable.
+  /// Called by MemoryManager and propagates down recursively if applicable.
   virtual void uncapMemoryAllocation() = 0;
-  // We might need to freeze memory allocating operations under severe global
-  // memory pressure.
+  /// We might need to freeze memory allocating operations under severe global
+  /// memory pressure.
   virtual bool isMemoryCapped() const = 0;
 
-  ////////////////////////// tree methods ////////////////////////////
-  // TODO: might be able to move implementation to abstract class
-  // name of the node, preferably unique among siblings, but not enforced.
-  virtual const std::string& getName() const = 0;
-  virtual std::weak_ptr<MemoryPool> getWeakPtr() = 0;
-  virtual MemoryPool& getChildByName(const std::string& name) = 0;
-  virtual uint64_t getChildCount() const = 0;
-  virtual void visitChildren(
-      std::function<void(MemoryPool* FOLLY_NONNULL)> visitor) const = 0;
-
-  virtual std::shared_ptr<MemoryPool> genChild(
-      std::weak_ptr<MemoryPool> parent,
-      const std::string& name,
-      int64_t cap) = 0;
-  virtual MemoryPool& addChild(
-      const std::string& name,
-      int64_t cap = kMaxMemory) = 0;
-  virtual std::unique_ptr<ScopedMemoryPool> addScopedChild(
-      const std::string& name,
-      int64_t cap = kMaxMemory) = 0;
-  virtual void dropChild(const MemoryPool* FOLLY_NONNULL child) = 0;
-
-  // Used to explicitly remove self when a user is done with the node, since
-  // nodes are owned by parents and users only get object references. Any
-  // reference to self would be invalid after this call. This will also drop the
-  // entire subtree.
-  virtual void removeSelf() = 0;
-
-  // Used to manage existing externally allocated memories without doing a new
-  // allocation.
+  /// Used to manage existing externally allocated memories without doing a new
+  /// allocation.
   virtual void reserve(int64_t /* bytes */) {
     VELOX_NYI("reserve() needs to be implemented in derived memory pool.");
   }
-  // Sometimes in memory governance we want to mock an update for quota
-  // accounting purposes and different implementations can
-  // choose to accommodate this differently.
+  /// Sometimes in memory governance we want to mock an update for quota
+  /// accounting purposes and different implementations can
+  /// choose to accommodate this differently.
   virtual void release(int64_t /* bytes */, bool /* mock */ = false) {
     VELOX_NYI("release() needs to be implemented in derived memory pool.");
   }
+
+ protected:
+  virtual std::shared_ptr<MemoryPool> genChild(
+      std::shared_ptr<MemoryPool> parent,
+      const std::string& name,
+      int64_t cap) = 0;
+
+  /// Drops child (and implicitly the entire subtree) to reclaim memory.
```

Comment:
Perhaps, explain how this is used, i.e. what if there are still live buffers and vectors allocated from the memory pool.

Reply:
Updated. This has nothing with memory reclamation. This is called on pool destruction so all its allocated buffer should have been freed.